### PR TITLE
LazyMap: prevent value building

### LINKED ISSUE #375: LazyMap has to prevent Value building
### Expected Behavior

The Method Value() in a self implemented IKvp is a method which shall not be called from constructors. In order to implement `IEnumerable<KeyValuePair<>>` from LazyDict the GetEnumerable() Method has to be implemented.> It has to return fully generated `KeyValuePair<>` Objects. In order to do that `Value()` has to be called.

When using a `LazyDict` within a constructor accepting `IEnumerable<KeyValuePair<>>` the `.GetEnumerator()` will be called which results in calling `Value()` which is not allowed in constructors.

An Exception has to be thrown in order to prevent this from happening.

As this might be the wanted behaviour it needs to be enabled or disabled by the user.

### Actual Behavior

No exception is thrown, `Value()` can be accidentally called from constructors.

### Steps to reproduce the behavior

```csharp
var dict = new LazyDict<int, int>(
    new FailingValueKvp(1)
);

var ex = Assert.Throws<Exception>(() => new List<KeyValuePair<int, int>>(dict));
Assert.Equal("i shall not be called", ex.Message);
```
with
```csharp
        private sealed class FailingValueKvp : IKvp<int, int>
        {
            private readonly int key;
            public FailingValueKvp(int key)
            {
                this.key = key;
            }

            public int Value()
            {
                throw new Exception("i shall not be called");
            }

            public int Key()
            {
                return key;
            }
        }
```

### LINKED ISSUE #377: IKvp should know if its lazy
### Expected Behavior
IKvp has a method bool IsLazy() which tells, if its value is built lazily.
With this knowledge, LazyMap can look at its kvps and can decide if it needs to reject getting an enumerator or not.

The is done inside a sticky scalar.

### Actual Behavior
IKvp is undefined about being lazy or not from the outside.

### PR DESCRIPTION
**Pull request by EdJoPaTo**
### Please check if the PR fulfills these requirements
- [x] I made sure that my code builds
- [x] I merged the master into this branch before pushing
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?** (check one with "x")
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

### What is the current behavior? (You can also link to an open issue here)

closes #375
closes #377

### Does this PR introduce a breaking change? (check one with "x")
- [x] Yes
- [ ] No

### If this PR contains a breaking change, please describe the impact and migration path for existing applications

The LazyMap does not build all the elements anymore with the same arguments. `rejectBuildingAllValues` has to be set to false in order to use it like before.

### CODE REVIEW

**Comment by EdJoPaTo** [comment]
tests currently dont run as other classes depend on LazyMap in some way. Im not sure how to fix that as that requires deeper understanding of the affected classes, their goal and their implementation specifics

**Comment by EdJoPaTo** [comment]
Include logic of #377 in order to prevent as many exceptions in cases where values were already built.

File: src/Yaapii.Atoms/Map/LazyDict.cs [hunk 1]
```diff
@@ -60,36 +69,60 @@ public LazyDict(IEnumerable<IKvp> kvps)
                     }
                     return dict;
                 });
+            this.anyValueIsLazy = new Sticky<bool>(() =>
+            {
+                return new Ternary<IFail, bool>(
```
Comment (Meerownymous) [inline src/Yaapii.Atoms/Map/LazyDict.cs]: It seems easier to me if we combine the mapped with a Filtered and ask its length.

But for performance reasons, I suggest we simply move an enumerator and break if we find a lazy kvp. That way we reduce the test to one iteration which we also might not need to fully finish.

File: tests/Yaapii.Atoms.Tests/Map/SortedTest.cs [hunk 2]
```diff
@@ -104,8 +104,12 @@ public void DoesNotBuildValueWhenNotNeeded()
             Assert.Equal(1, keys[1]);
             Assert.Equal(6, keys[2]);
 
-            var ex = Assert.Throws<Exception>(() => sorted.Values.GetEnumerator());
-            Assert.Equal("i shall not be called", ex.Message);
+            // Sanety check: building one value isn't a problem
+            Assert.Equal(4, sorted[1]);
```
Comment (Meerownymous) [inline tests/Yaapii.Atoms.Tests/Map/SortedTest.cs]: we have the rule that we make only one assert per test. Please refactor.
Please remove comments (the method names of the tests, which are facts, will tell us what your comment currently does)

File: tests/Yaapii.Atoms.Tests/Map/LazyDictTests.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        public void CanGetEnumeratorWhenEmpty()
+        {
+            var dict = new LazyDict<int, int>(true);
+            dict.GetEnumerator();
+        }
+
+        /// <summary>
+        /// IKvp<int, int> with an always failing Value()
+        /// </summary>
+        private sealed class FailingValueKvp : IKvp<int, int>
```
Comment (Meerownymous) [inline tests/Yaapii.Atoms.Tests/Map/LazyDictTests.cs]: pleae use a FkKvp to which you pass a function. Remove inline class from test class.

File: tests/Yaapii.Atoms.Tests/Map/SortedTest.cs [hunk 4]
```diff
@@ -37,77 +36,84 @@ public void ValueStillBehindCorrectKeyIEnumerableKeyValuePairs(int key, int expe
...
+                new Kvp.Of<int, int>(6, () => { throw new Exception("i shall not be called"); }),
+                new Kvp.Of<int, int>(-5, () => { throw new Exception("i shall not be called"); })
+            );
+            var sorted = new Sorted<int, int>(unsorted);
             Assert.Equal(4, sorted[1]);
+        }
 
-            // Sanety check: but all of them is
+        [Fact]
+        public void BuildingAllValuesThrowsWithEnabledBuildAllRejection()
```
Comment (Meerownymous) [inline tests/Yaapii.Atoms.Tests/Map/SortedTest.cs]:
```suggestion
        public void RejectsBuildingAllValuesByDefault()
```

File: tests/Yaapii.Atoms.Tests/Map/SortedTest.cs [hunk 5]
```diff
@@ -37,77 +36,84 @@ public void ValueStillBehindCorrectKeyIEnumerableKeyValuePairs(int key, int expe
...
+            Assert.Equal(expectedKey, keys[index]);
         }
 
-        [Fact]
-        public void DoesNotBuildValueWhenNotNeeded()
+        [Theory]
+        [InlineData(0, -5)]
+        [InlineData(1, 1)]
+        [InlineData(2, 6)]
+        public void DoesNotBuildValueWhenNotNeeded(int index, int expectedKey)
```
Comment (Meerownymous) [inline tests/Yaapii.Atoms.Tests/Map/SortedTest.cs]:
```suggestion
        public void EnumeratesKeysWhenLazy(int index, int expectedKey)
```

File: tests/Yaapii.Atoms.Tests/Map/SortedTest.cs [hunk 6]
```diff
@@ -37,77 +36,84 @@ public void ValueStillBehindCorrectKeyIEnumerableKeyValuePairs(int key, int expe
...
+            Assert.Equal(expectedKey, keys[index]);
+        }
 
-            Assert.Equal(-5, keys[0]);
-            Assert.Equal(1, keys[1]);
-            Assert.Equal(6, keys[2]);
-
-            // Sanety check: building one value isn't a problem
+        [Fact]
+        public void BuildingSingleValueIsStillOkWithBuildAllRejection()
```
Comment (Meerownymous) [inline tests/Yaapii.Atoms.Tests/Map/SortedTest.cs]:
```suggestion
        public void DeliversSingleValueWhenLazy()
```

**Pull request merged**